clifford/_mvarray.py

```diff
@@ -1,57 +1,96 @@
-import numpy as np
+import functools
+import operator
+from typing import Tuple
 
+import numpy as np
 
 from clifford.io import write_ga_file, read_ga_file  # noqa: F401
+from ._layout import Layout
 from ._multivector import MultiVector
 
 dual_array = np.vectorize(MultiVector.dual)
 normal_array = np.vectorize(MultiVector.normal)
 call_array = np.vectorize(MultiVector.__call__)
 
 
+def _interrogate_nested_mvs(input_array) -> Tuple[Tuple[int], Layout, np.dtype]:
+    """
+    Calculates the shape of the nested input_array, and gets the associated layout.
+    Stops descending when it encounters a MultiVector.
+    """
+    if not isinstance(input_array, MultiVector):
+        nested_shape, layout, dtype = _interrogate_nested_mvs(input_array[0])
+        return (len(input_array), *nested_shape), layout, dtype
+    else:
+        return (), input_array.layout, input_array.value.dtype
+
+
+def _index_nested_iterable(input_iterable, index):
+    """
+    Given a nested iterable, input_iterable, return the element given by the
+    1d index iterable
+    """
+    return functools.reduce(operator.getitem, index, input_iterable)
+
+
+@functools.lru_cache
```

Comment:
```suggestion
@functools.lru_cache(None)
```
Needs an argument for the cache size